### PR TITLE
Added initial support for tusd

### PR DESCRIPTION
**Pull request by nuwang**
This builds on @mvdbeek's fantastic work on integrating tus with Galaxy: https://github.com/galaxyproject/galaxy/pull/12656, by addingthe tusd daemon as an out-of-the-box component in the helm chart.

Things are at a point where tusd is receiving the requests, but something is slightly amiss with auth handling:

```
[tusd] 2021/10/12 14:06:23 Using '/galaxy/server/database/tmp' as directory storage.
[tusd] 2021/10/12 14:06:23 Using 0.00MB as maximum size.
[tusd] 2021/10/12 14:06:23 Using 'http://gxyarchivetest6-galaxy-nginx:8000/galaxy/api/upload/hooks' as the endpoint for hooks
[tusd] 2021/10/12 14:06:23 Enabled hook events: pre-create, post-create, post-receive, post-terminate, post-finish
[tusd] 2021/10/12 14:06:23 Using 0.0.0.0:1080 as address to listen.
[tusd] 2021/10/12 14:06:23 Using /files/ as the base path.
[tusd] 2021/10/12 14:06:23 Using /metrics as the metrics path.
[tusd] 2021/10/12 14:06:23 Supported tus extensions: creation,creation-with-upload,termination,concatenation,creation-defer-length
[tusd] 2021/10/12 14:06:23 You can now upload files to: http://0.0.0.0:1080/files/
[tusd] 2021/10/12 14:11:34 event="RequestIncoming" method="POST" path="" requestId="76a08832e52cef184ed906db1d5c4510"
[tusd] 2021/10/12 14:11:34 event="HookInvocationStart" type="pre-create" id=""
[tusd] 2021/10/12 14:11:34 event="ResponseOutgoing" status="403" method="POST" path="" error="pre-create hook failed: endpoint returned: 403 Forbidden" requestId="76a08832e52cef184ed906db1d5c4510"
```

session cookie not getting forwarded perhaps? Ping @mvdbeek

### CODE REVIEW

File: galaxy/templates/deployment-tusd.yaml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          image: "{{ .Values.tusd.image.repository }}:{{ .Values.tusd.image.tag }}"
+          imagePullPolicy: {{ .Values.tusd.image.pullPolicy }}
+          args:
+            - "-host"
+            - "0.0.0.0"
+            - "-port"
+            - "1080"
+            - "-upload-dir={{ .Values.persistence.mountPath }}/tmp"
+            - '-hooks-http=http://{{ include "galaxy.fullname" . }}-nginx:{{ .Values.service.port }}{{ template "galaxy.add_trailing_slash" .Values.ingress.path }}api/upload/hooks'
+            - "-hooks-http-forward-headers=X-Api-Key,sessioncookie"
```
Comment (mvdbeek) [inline galaxy/templates/deployment-tusd.yaml:48]:
```suggestion
            - "-hooks-http-forward-headers=X-Api-Key,Cookie"
```

File: galaxy/templates/configmap-nginx.yaml [hunk 2]
```diff
@@ -69,6 +69,19 @@ data:
...
+                # Disable request and response buffering
+                proxy_request_buffering  off;
+                proxy_buffering          off;
+                proxy_http_version       1.1;
+                # Add X-Forwarded-* headers
+                proxy_set_header X-Forwarded-Host $host;
+                proxy_set_header X-Forwarded-Proto $scheme;
+                proxy_set_header         Upgrade $http_upgrade;
+                proxy_set_header         Connection "upgrade";
+                client_max_body_size     {{ .Values.nginx.conf.client_max_body_size }};
```
Comment (mvdbeek) [inline galaxy/templates/configmap-nginx.yaml:82]: I think this should be 0, a well-behaved client shouldn't send a body to api/upload/resumable_upload

**Comment by nuwang** [comment]
@mvdbeek The `Cookie` header solved the auth issue, thanks. But there's a small issue remaining, with Galaxy complaining that:

image

```
Uploaded temporary file (/galaxy/server/database/tmp/resumable_upload7b8816eaa9268e559a64702b63173feb) does not exist.
```

The content of the upload dir does not have resumable_upload prefixed in front.

```
/srv/tusd-data $ ls -alh /galaxy/server/database/tmp
total 140
drwxr-xr--    2 101      101         4.0K Oct 12 14:41 .
drwxrwxrwx   13 root     root        4.0K Oct 12 14:41 ..
-rw-r--r--    1 101      101       124.2K Oct 12 14:40 7b8816eaa9268e559a64702b63173feb
-rw-r--r--    1 101      101          570 Oct 12 14:40 7b8816eaa9268e559a64702b63173feb.info
``

File: galaxy/templates/configmap-nginx.yaml [hunk 3]
```diff
@@ -69,6 +69,19 @@ data:
...
+                proxy_pass http://{{ template "galaxy.fullname" . }}-tusd:1080/files/;
```
Comment (mvdbeek) [inline galaxy/templates/configmap-nginx.yaml:83]:
```suggestion
                proxy_pass http://{{ template "galaxy.fullname" . }}-tusd:1080/files;
```
I also ran into the same thing and then didn't update the docs ...

**Comment by nuwang** [comment]
For whatever reason, it doesn't work if I log out, it's using the legacy upload:
image

If I log back in, I still can't upload that failed file. This is after a hard refresh.

**Comment by mvdbeek** [comment]
Hmm, in both cases you have a sessioncookie, so that's all the formal requirements. But I see we do some additional logic that might disable chunked uploads. https://github.com/galaxyproject/galaxy/pull/12656/commits/5f926d966e1a84f9bca9cae3c962db3c42b90238 might fix this (I haven't tested it yet)

**Comment by mvdbeek** [comment]
> * As shown in the last screen show above, all of the HEAD requests return 200, before finally logging this:
>   `Failed because: Error: tus: invalid or missing offset value, originated from request (method: HEAD, url: http://localhost/galaxy/api/upload/resumable_uploadf64e9c47874b96195d41b1f4d983b663, response code: 200, response text: , request id: n/a)`
>   It also seems to be issuing the same HEAD request over and over again.

Yeah, I also noticed this just now. It seems to be quote selective about the file in question, which is weird. That is probably a client or nginx problem.

**Review by mvdbeek — APPROVED** [review]

**Review by afgane — APPROVED** [review]

**Pull request merged**